test/infrastructure/docker/docker/kind_manager.go

```diff
@@ -17,16 +17,22 @@ limitations under the License.
 package docker
 
 import (
+	"context"
 	"fmt"
 	"net"
 	"os"
 	"strings"
 
+	dockerTypes "github.com/docker/docker/api/types"
+	dockerContainer "github.com/docker/docker/api/types/container"
+	dockerMount "github.com/docker/docker/api/types/mount"
+	dockerNetwork "github.com/docker/docker/api/types/network"
+	dockerClient "github.com/docker/docker/client"
+	"github.com/docker/go-connections/nat"
```

Comment:
I'm thinking a few phases to get there to make it easier to review and to minimize risk from changing too much.

Step 1 is this of just getting rid of the exec calls and sorting out the proper API calls to make in their place.
Step 2 would be to do that larger refactoring of code to structure things in a way that would allow plugging in fakes or other CRI providers.
Step 3 (or step 2b) would then be where we should be able fully implement unit test coverage by faking out the API calls through that cleaner interface.

Reply:
What do you think, should there be an issue capturing the "big picture"? I think https://github.com/kubernetes-sigs/cluster-api/issues/4380 only covers step 1 at the moment.

Reply:
+1 to make the  "big picture" more transparent; up to you use one or more issues

Reply:
I wonder if these next steps warrant a CAEP? Since we want to abstract the docker interaction behind an interface, that also opens up the possibility to be able to plug in other runtimes behind that interface to support non-Docker deployments. Should that be in scope so we design that interface with that in mind?

Or would that be too much for now, and I should just stick to refactoring the docker engine interaction and we can decide at a future date to tweak whatever we need if/when we want to expand beyond that?

Reply:
I don't have strong preference for a CAEP, given that this is mostly an internal refactor/not user facing.

However, I think that the "fake CRI" we should use for testing falls into "abstract away docker behind an interface", so we can start working on this and then introduce other CRIs later (the only use case that comes to mind mind for using containerd instead of docker is when running CAPD on CI, this is not really an high priority given that we are already setup for using docker)

Reply:
Updated the issue to call out the separate steps in the plan. Updated this PR so it does not close the issue, just link to it.